src/librustc_mir/transform/check_consts/qualifs.rs

```diff
@@ -113,7 +113,9 @@ where
     F: FnMut(Local) -> bool,
 {
     match rvalue {
-        Rvalue::NullaryOp(..) => Q::in_any_value_of_ty(cx, rvalue.ty(*cx.body, cx.tcx)),
+        Rvalue::Tls(_) | Rvalue::NullaryOp(..) => {
+            Q::in_any_value_of_ty(cx, rvalue.ty(*cx.body, cx.tcx))
+        }
```

Comment:
Shouldn't `Tls` be outright rejected?

Reply:
Depends on the qualif I guess? I mean before this PR we used to get the qualif from the constant pointer, so this change replicates the previous behaviour.

Reply:
I don't know what "get the qualif from the constant pointer" means, but you said before that you used `LocalInfo`, and then `Rvalue::Tls` seems like a much better signal (ideally together with an assertion somewhere that constant pointers we see elsewhere do *not* point to a thread-local).

Reply:
Previously what is now `Rvalue::Tls` was `Rvalue::Use(Operand::Constant(...))` and the qualification was obtained from the type of the constant, so there was no special treatment of thread locals. There is no way for qualifications to report errors, the only signal is whether the qualifcation is there or not. Since we don't know what `true` or `false` means for the concrete qualification, the generic qualification logic can't make any assumptions here.

I guess there are some assumptions we do, so I guess I can return `false` here. @ecstatic-morse does that seem right?

Reply:
Oh so this was not where TLS accesses were rejected. Sorry for the misunderstanding.

But somewhere we should make sure that MIR containing `Rvalue::Tls` is const-rejected. Where is that happening?

Reply:
> you said before that you used LocalInfo, and then Rvalue::Tls seems like a much better signal (ideally together with an assertion somewhere that constant pointers we see elsewhere do not point to a thread-local).

That's not here. What you're looking for is the const checking logic in https://github.com/rust-lang/rust/blob/0a400bde2a7d2c0d4c9e695ee7bbf2d431e528fa/src/librustc_mir/transform/check_consts/validation.rs#L438 which is a different thing from qualification (and that doesn't even use the debuginfo for thread locals, only for statics in general).

Reply:
You've probably seen this by now, but cross linking anyway. `Rvalue::Tls` is const-rejected in https://github.com/rust-lang/rust/pull/71192#discussion_r409649156

Reply:
Makes sense, thanks!